applications/tari_base_node/src/grpc.rs

```diff
@@ -0,0 +1,221 @@
+    blocks::BlockHeader,
+    mempool::service::LocalMempoolService,
+    proof_of_work::PowAlgorithm,
+};
+use tari_crypto::tari_utilities::{epoch_time::EpochTime, ByteArray, Hashable};
+use tokio::runtime;
+use tonic::{transport::Server, Code, Request, Response, Status};
+
+const LOG_TARGET: &str = "base_node::grpc";
+
+pub mod base_node_grpc {
+    tonic::include_proto!("tari.base_node");
+}
+
+pub struct BaseNodeGrpcServer {
+    executor: runtime::Handle,
+    node_service: LocalNodeCommsInterface,
+    mempool_service: LocalMempoolService,
+}
+
+impl BaseNodeGrpcServer {
+    pub fn new(executor: runtime::Handle, ctx: &NodeContainer) -> Self {
+        Self {
+            executor,
+            node_service: ctx.local_node(),
+            mempool_service: ctx.local_mempool(),
+        }
+    }
+}
+
+#[tonic::async_trait]
+impl base_node_grpc::base_node_server::BaseNode for BaseNodeGrpcServer {
+    async fn list_headers(
+        &self,
+        request: Request<ListHeadersRequest>,
+    ) -> Result<Response<ListHeadersResponse>, Status>
+    {
+        trace!(target: LOG_TARGET, "List headers called:{:?}", request);
+        let request = request.into_inner();
+        let mut handler = self.node_service.clone();
+        let tip = match handler.get_metadata().await {
+            Err(err) => {
+                warn!(target: LOG_TARGET, "Error communicating with base node: {}", err,);
+                return Err(Status::internal(err.to_string()));
+            },
+            Ok(data) => data.height_of_longest_chain.unwrap_or(0),
+        };
+
+        let sorting: Sorting = request.sorting();
+        let num_headers = match request.num_headers {
+            0 => 10,
+            _ => request.num_headers,
+        };
+
+        let headers: Vec<u64> = if request.from_height != 0 {
+            match sorting {
+                Sorting::Desc => ((request.from_height - num_headers + 1)..=request.from_height).collect(),
+                Sorting::Asc => (request.from_height..(request.from_height + num_headers)).collect(),
+            }
+        } else {
+            match sorting {
+                Sorting::Desc => ((tip - num_headers)..tip).collect(),
+                Sorting::Asc => (0..num_headers).collect(),
+            }
+        };
+
+        let mut headers = match handler.get_headers(headers).await {
+            Err(err) => {
+                warn!(target: LOG_TARGET, "Error communicating with base node: {}", err,);
+                return Err(Status::internal(err.to_string()));
+            },
+            Ok(data) => data,
+        };
+        headers.sort_by(|a, b| match sorting {
+            Sorting::Desc => b.height.cmp(&a.height),
+            Sorting::Asc => a.height.cmp(&b.height),
+        });
+        let reply = ListHeadersResponse {
+            headers: headers.iter().map(|h| convert_block_header(h)).collect(),
+        };
+
+        debug!(target: LOG_TARGET, "Responding:{:?}", reply);
```

Comment:
Debug messages must be single line please
```suggestion
        debug!(target: LOG_TARGET, "ListHeaders response: {...}", reply);
```